src/parsers.rs

```diff
@@ -115,6 +119,96 @@ impl<'a> Parser<'a> {
         Err(())
     }
 
+    #[cfg(feature = "ipv6")]
+    fn accept_ipv6(&mut self, is_cidr: bool) -> Result<Ipv6Address> {
```

Comment:
I tried to simplify this as much as possible, but since it is of a variable length, parsing is a bit complex. Suggestions on methods to make this more readable are very much welcomed.

Reply:
The easiest way to make this parser much simpler would be to make it recursive. Here's an outline:
```
fn accept_ipv6_part(&mut self, 
      (head, tail): (&mut [u16; 8], &mut [u16; 8]), 
      (head_idx, tail_idx): (&mut usize, &mut usize),
      mut use_tail: bool) -> Result<()> {
  match self.accept_str(b"::") {
    Ok(()) => use_tail = true,
    Err(()) => ()
  }

  if head_idx != 0 || use_tail && tail_idx != 0 {
    self.accept_char(b':')?;
  }

  match self.accept_number(4, 0x10000, true)? {
    Ok(part) if use_tail => {
      tail[tail_idx] = part;
      tail_idx += 1;
    }
    Ok(part) => {
      head[head_idx] = part;
      head_idx += 1;        
    }
    Err(()) => ()
  }

  if head_idx + tail_idx > 8 {
    Err(())
  } else if self.pos == self.data.len() {
    head[8 - tail.len()..].copy_from_slice(tail);
    Ok(())
  } else {
    self.accept_ipv6_part((head, tail), (head_idx, tail_idx), use_tail)
  }
}
```

Reply:
Can we rely on `(self.pos == self.data.len()) == true` event for a CIDR? I thought for the current implementation I just had to hand processing over as soon as I hit a non-number in the second parsing round or when I filled the addr buffer. Either way, adding that to the above would be pretty easy and this is def more readable. I'll work on this tomorrow after I get some sleep. I need more brain power to think through this :smile:

Reply:
Oh I guess you can look ahead for whether there's a `:` and bail out if there is not, instead of looking for EOF.

Reply:
Updated to use the recursive method and used `lookahead_char` instead of `is_cidr`. IMO it makes it less confusing and there are less mutable variables you need to keep track of.